arch/arm/core/aarch32/thread.c

```diff
@@ -420,3 +430,48 @@ void arch_switch_to_main_thread(struct k_thread *main_thread, char *stack_ptr,
 
 	CODE_UNREACHABLE;
 }
+
+#if !defined(CONFIG_MULTITHREADING) && defined(CONFIG_CPU_CORTEX_M)
+
+extern K_THREAD_STACK_DEFINE(z_main_stack, CONFIG_MAIN_STACK_SIZE);
+
+FUNC_NORETURN void z_arm_switch_to_main_no_multithreading(
+	k_thread_entry_t main_entry, void *p1, void *p2, void *p3)
+{
+	z_arm_prepare_switch_to_main();
+
+	/* Set PSP to the highest address of the main stack. */
+	char *psp =	Z_THREAD_STACK_BUFFER(z_main_stack) +
+		K_THREAD_STACK_SIZEOF(z_main_stack);
+
+#if defined(CONFIG_BUILTIN_STACK_GUARD)
+	char *psplim = (Z_THREAD_STACK_BUFFER(z_main_stack));
+	/* Clear PSPLIM before setting it to guard the main stack area. */
+	__set_PSPLIM(0);
+#endif
+
+	__asm__ volatile (
+	"mov  r0, %0\n\t"
+	"mov  r1, %1\n\t"
+	"mov  r2, %2\n\t"
```

Comment:
Is it not possible to use local register variables here instead, I mean:
```
	register void* p1_input __asm__ ("r0") = p1;
	register void* p2_input __asm__ ("r1") = p2;
	register void* p3_input __asm__ ("r2") = p3;
	__asm__ volatile (
#ifdef CONFIG_BUILTIN_STACK_GUARD
	"msr  PSPLIM, %5\n\t"
#endif
	"msr  PSP, %3\n\t"
	"blx  %4\n\t" /* call main entry hook */
	:
	: "r" (p1_input), "r" (p2_input), "r" (p3_input),
	  "r" (psp), "r" (main_entry)
#ifdef CONFIG_BUILTIN_STACK_GUARD
	  , "r" (psplim)
#endif
	);
```
and avoid specifying those registers as clobbers?

Reply:
Sort of a matter of taste, no? I reduce to clobber list to r0-r2. Like this, it is similar to the function arch_switch_to_main_thread, in this file.

Reply:
Well, it's not solely a matter of taste. The way it is now, the compiler is forced to use some other registers just for preparing values to be written to r0-r2 in the inline assembly code. Because of the above call to `z_arm_prepare_switch_to_main()`, the compiler anyway needs to store somewhere (probably in some higher registers) parameters p1-p3 that were passed in registers r1-r3, so in this particular case there will be probably no noticeable difference. By using local register variables, you make the compiler aware of what you in fact need for the inline assembler and you let it do the best register allocation it can. It could matter in some other cases for which this piece of code could be copied in some future, or if this function gets modified at some point.